kpi/views.py

```diff
@@ -748,15 +749,24 @@ def create(self, request, *args, **kwargs):
             asset_uid = self.get_parents_query_dict().get("asset")
             asset = get_object_or_404(self.parent_model, uid=asset_uid)
             instance_id = request.data.get("instance_id")
-            if not HookUtils.call_services(asset, instance_id):
+            instance = asset.deployment.get_submission(instance_id)
+
+            # Check if instance really belongs to Asset.
+            if not (instance and instance.get(asset.deployment.INSTANCE_ID_FIELDNAME) == instance_id):
+                response_status_code = status.HTTP_404_NOT_FOUND
+                response = {
+                    "detail": _("Resource not found")
+                }
+
+            elif not HookUtils.call_services(asset, instance_id):
                 response_status_code = status.HTTP_409_CONFLICT
                 response = {
                     "detail": _(
                         "Your data for instance {} has been already submitted.".format(instance_id))
                 }
 
         except Exception as e:
```

Comment:
This should be changed to a more specific type of exception, and the `try` should probably include only `call_services()`. If it has to stay `except Exception as e`, then we should use `six.reraise()` or something to keep the original exception information while adding a descriptive message